### PR TITLE
Addition of lenses for PParams

### LINKED ISSUE #2927: Switch `PParams` to use lenses instead of `HasField`
In a recent PR #2901 we've switched all transaction related types to use lenses. Same needs to be done for PParams. 

Also concrete types need to become newtypes around the `*PParamHKD`, rather than the type synonyms. Also renaming fields using proper convention could be done at the same time. 

- [ ] - Use `newtypes` instead of `type` synonyms
- [ ] - Define lenses and provide instances
- [ ] - Rename fields

So the end goal should look something along the lines:

* The type class:
```haskell
class (...) => EraPParams era where
  -- | Protocol parameters
  type PParams era = (r :: Type) | r -> era

  -- | The type of updates to Protocol parameters
  type PParamsUpdate era = (r :: Type) | r -> era

  emptyPParams :: PParams era
  emptyPParamsUpdate:: PParamsUpdate era

  applyPPUpdates :: PParams era -> PParamsUpdate era -> PParams era

  ppMinFeeA_L :: Lens' (PParams (ShelleyEra crypto)) Natural
  ppMinfeeB_L :: Lens' (PParams (ShelleyEra crypto)) Natural
  ppMaxBBSize_L :: Lens' (PParams (ShelleyEra crypto)) Natural
  ...
  ppA0_L :: Lens' (PParams (ShelleyEra crypto)) NonNegativeInterval
  ppRho_L :: Lens' (PParams (ShelleyEra crypto)) UnitInterval
  ppTau_L :: Lens' (PParams (ShelleyEra crypto))  UnitInterval
  ppD_L :: Lens' (PParams (ShelleyEra crypto))  UnitInterval

  ppuMinfeeA_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe Natural)
  ppuMinfeeB_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe Natural)
  ppuMaxBBSize_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe Natural)
  ...
  ppuA0_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe NonNegativeInterval)
  ppuRho_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe UnitInterval)
  ppuTau_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe UnitInterval)
  ppuD_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe UnitInterval)
```

I am not set on the names (for example I am not fond of underscore), but this is the best I could come up with.

Then instances can be defined as:
```haskell
data ShelleyPParamsHKD f era = ShelleyPParams
  { -- | The linear factor for the minimum fee calculation
    sppMinFeeA :: !(HKD f Natural),
  ...
  }

sppMinFeeA_L :: Lens' (ShelleyPParamsHKD f (ShelleyEra crypto)) (HKD f Natural),
sppMinFeeA_L = lens sppMinFeeA (\pp minFeeA -> pp { sppMinFeeA = minFeeA })

newtype ShelleyPParams era = ShelleyPParams (ShelleyPParamsHKD Identity era)

newtype ShelleyPParamsUpdate era = ShelleyPParamsUpdate (ShelleyPParamsHKD StrictMaybe era)

instance CC.Crypto crypto => EraPParams (ShelleyEra crypto) where
  type PParams (ShelleyEra crypto) = ShelleyPParams (ShelleyEra crypto)
  type PParamsUpdate (ShelleyEra crypto) = ShelleyPParamsUpdate (ShelleyEra crypto)

  applyPPUpdates = updatePParams

  ppMinFeeA_L = coerce sppMinFeeA_L
  ...
  ppuMinFeeA_L = coerce sppMinFeeA_L
  ...
```

### PR DESCRIPTION
**Pull request by lehins**
# Description

Addition of  `EraPParams`, `AlonzoEraPParams` and `BabageEraPParams` type classes with lenses that can access and manipulate both `PParams` and `ParamsUpdate`. Each type class requires implementation lenses for the `PParamsHKD` type, which are later used for defining specific lenses for PParams and PParamsUpdate respectfully

HKD types `ShelleyPParams`, `AlonzoPParams` and `BabbagePParams` now become internal, for the most part and the `PParams`+`PParamsUpdate` are the newtype wrappers around the era specific internal HKD type. All operations on PParams should be done with lenses from now on.

Other notable things in this PR:

* Fixes `Ord` instances for `AlonzoPParams` and `BabbagePParams` by introducing an internal `OrdExUniits type.
* Moves around `To/FromJSON` instances to prevent the from being orphan
* Moved `CostModels` into `Cardano.Ledger.Alozno.Scripts` from `Cardano.Ledger.Alozno.PParams`
* Switch all functions that accepted PParams as an argument, but only used protocol version, to accept only protocol version. Most notable examples are functions that do conditional checking on the hardfork
* Makes `PoolEnv` fields strict
* Adds a data type `ByronEra` that has an `Era` instance. This is done mostly for completeness, but also to allow getting protocol version from the type level. Also this allowed `PreviousEra` type family to be moved into `Era` type class.
* Moved all class definitions {EraName}Era{Type}, into the `Cardano.Ledger.{EraName}.Core` module. It was discussed before, but wasn't fully implemented.

Fixes #2927 

# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the [changelog](https://github.com/input-output-hk/cardano-ledger/blob/master/CHANGELOG.md)
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (which can be run with `scripts/fourmolize.sh`
- [x] Self-reviewed the diff

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,9 @@ in the naming of release branches.
 
 ### Added
 
+- Added type classes: `EraPParams`, `AlonzoEraPParams` and `BabageEraPParams` with lenses
```
Comment (Soupstraw) [inline CHANGELOG.md]:
```suggestion
- Added type classes: `EraPParams`, `AlonzoEraPParams` and `BabbageEraPParams` with lenses
```

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs [hunk 2]
```diff
@@ -112,165 +144,46 @@ instance FromCBOR AlonzoGenesis where
...
-        , collateralPercentage
-        , maxCollateralInputs
+        { agCoinsPerUTxOWord
+        , agCostModels
+        , agPrices
+        , agMaxTxExUnits
+        , agMaxBlockExUnits
+        , agMaxValSize
+        , agCollateralPercentage
+        , agMaxCollateralInputs
```
Comment (Soupstraw) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs]: This could be done more compactly with sequential application:
```haskell
instance FromJSON AlonzoGenesis where
  parseJSON = Aeson.withObject "Alonzo Genesis" $ \o ->
    AlonzoGenesis
      <$> o .: "lovelacePerUTxOWord"
      <*> o .: "costModels"
      <*> ...
```

File: CHANGELOG.md (hunk 1 above)
Comment (lehins) [reply CHANGELOG.md]: Done

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs (hunk 2 above)
Comment (lehins) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs]: I don't like this style cause it is easy to mess up the order. For example if the order of fields has changed in the type itself, or I just happen to mess up the order in the implementation then compiler is not gonna help me in any way:

```diff
instance FromJSON AlonzoGenesis where
  parseJSON = Aeson.withObject "Alonzo Genesis" $ \o ->
    AlonzoGenesis
      <$> o .: "lovelacePerUTxOWord"
      <*> o .: "costModels"
      <*> ...
-      <*> o .: "agMaxTxExUnits"
-      <*> o .: "agMaxBlockExUnits"
+      <*> o .: "agMaxBlockExUnits"
+      <*> o .: "agMaxTxExUnits"
      <*> ...
```


That being said, I could use the record wildcards at the end, which would make just as compact as the version above, but still safe:
```haskell
    return AlonzoGenesis {..}
```

I'll do that when there will be any more suggestions from others

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs [hunk 3]
```diff
@@ -669,7 +654,7 @@ txOutDataHash = strictMaybeToMaybe . getDataHashBabbageTxOut
 
 txOutScript :: BabbageTxOut era -> Maybe (Script era)
 txOutScript = strictMaybeToMaybe . getScriptBabbageTxOut
-{-# DEPRECATED txOutScript "In favor of `dataTxOutL` or `getScriptBabbageTxOut`" #-}
+{-# DEPRECATED txOutScript "In favor of `dataTxOutL` or `getScriptBa\bbageTxOut`" #-}
```
Comment (Soupstraw) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs]:
```suggestion
{-# DEPRECATED txOutScript "In favor of `dataTxOutL` or `getScriptBabbageTxOut`" #-}
```

File: eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs [hunk 4]
```diff
@@ -272,21 +267,19 @@ addTokens ::
   StrictSeq (TxOut era) ->
   Maybe (StrictSeq (TxOut era))
 addTokens proxy tooLittleLovelace pparams ts (txOut :<| os) =
-  let v = txOut ^. valueTxOutL
-   in if Val.coin v < scaledMinDeposit v (getField @"_minUTxOValue" pparams)
-        then addTokens proxy (txOut :<| tooLittleLovelace) pparams ts os
-        else Just $ tooLittleLovelace >< addValToTxOut @era (MaryValue 0 ts) txOut <| os
+  if txOut ^. coinTxOutL < getMinCoinTxOut pparams txOut
```
Comment (Soupstraw) [inline eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:270]: Nice, much more readable

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Core.hs [hunk 5]
```diff
@@ -21,7 +21,7 @@ import Cardano.Ledger.Core
 import Cardano.Ledger.Crypto (Crypto)
 import Cardano.Ledger.Shelley.Delegation.Certificates (DCert)
 import Cardano.Ledger.Shelley.Era (ShelleyEra)
-import Cardano.Ledger.Shelley.PParams (Update)
+import Cardano.Ledger.Shelley.PParams
```
Comment (Soupstraw) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Core.hs]: Unnecessary change?

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs (hunk 3 above)
Comment (lehins) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs]: Thank you. Fixed

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Core.hs (hunk 5 above)
Comment (lehins) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/Core.hs]: Yep. Fixed.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs (hunk 2 above)
Comment (lehins) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs]: Done. Thank you.

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/API/ByronTranslation.hs [hunk 6]
```diff
@@ -21,9 +21,9 @@ import qualified Cardano.Crypto.Hashing as Hashing
 import Cardano.Ledger.Address (fromBoostrapCompactAddress, isBootstrapRedeemer)
 import Cardano.Ledger.BaseTypes (BlocksMade (..), TxIx (..))
 import Cardano.Ledger.Coin (CompactForm (CompactCoin))
-import Cardano.Ledger.Core (bootAddrTxOutF)
+import Cardano.Ledger.Core
```
Comment (Soupstraw) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/API/ByronTranslation.hs:24]: Unnecessary change

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/API/ByronTranslation.hs [hunk 7]
```diff
@@ -21,9 +21,9 @@ import qualified Cardano.Crypto.Hashing as Hashing
...
 import Cardano.Ledger.Crypto (Crypto (ADDRHASH))
-import Cardano.Ledger.EpochBoundary (emptySnapShots)
+import Cardano.Ledger.EpochBoundary
```
Comment (Soupstraw) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/API/ByronTranslation.hs]: Unnecessary change

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs [hunk 8]
```diff
@@ -241,12 +245,12 @@ blockEx2 =
...
     (mkOCert @c (coreNodeKeysBySchedule @(ShelleyEra c) ppEx 50) 0 (KESPeriod 0))
 
-pulserEx2 :: forall c. (ExMock c, C.UsesPP (ShelleyEra c)) => PulsingRewUpdate c
+pulserEx2 :: forall c. (ExMock c) => PulsingRewUpdate c
 pulserEx2 = makePulser' expectedStEx1
 
 expectedStEx2 ::
   forall c.
-  (ExMock c, C.UsesPP (ShelleyEra c)) =>
+  (ExMock c) =>
```
Comment (Soupstraw) [inline eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs]:
```suggestion
  ExMock c =>
```

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs [hunk 9]
```diff
@@ -259,7 +263,7 @@ expectedStEx2 =
 --
 -- Submit an empty block to trigger adopting the genesis delegation.
 genesisDelegation2 ::
-  (ExMock c, C.UsesPP (ShelleyEra c)) =>
+  (ExMock c) =>
```
Comment (Soupstraw) [inline eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs]:
```suggestion
  ExMock c =>
```

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/Mir.hs [hunk 10]
```diff
@@ -149,15 +148,15 @@ txbodyEx1 pot =
     SNothing
   where
     aliceInitCoin = Val.inject $ Coin $ 10 * 1000 * 1000 * 1000 * 1000 * 1000
-    aliceCoinEx1 = aliceInitCoin <-> (Val.inject $ feeTx1 <+> _keyDeposit ppEx)
+    aliceCoinEx1 = aliceInitCoin <-> (Val.inject $ feeTx1 <+> (Coin 7))
```
Comment (Soupstraw) [inline eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/Mir.hs]:
```suggestion
    aliceCoinEx1 = aliceInitCoin <-> (Val.inject $ feeTx1 <+> Coin 7)
```

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/MirTransfer.hs [hunk 11]
```diff
@@ -26,14 +26,15 @@ import Cardano.Ledger.Shelley.API (
   Ptr (..),
   ShelleyDELEG,
  )
-import Cardano.Ledger.Shelley.PParams (ShelleyPParamsHKD (..), emptyPParams)
+import Cardano.Ledger.Shelley.Core
```
Comment (Soupstraw) [inline eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/MirTransfer.hs:29]: Redundant import?

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs [hunk 12]
```diff
@@ -59,16 +61,11 @@ import Test.Cardano.Ledger.Shelley.MultiSigExamples (
...
 import Test.Cardano.Ledger.Shelley.Serialisation.Generators ()
-
--- Generators is needed for Arbitrary ShelleyPParams
-import Test.Cardano.Ledger.Shelley.Utils (applySTSTest, runShelleyBase, slotFromEpoch)
-import Test.QuickCheck (Property, discard, (===))
+import Test.Cardano.Ledger.Shelley.Utils
 import Test.Tasty (TestTree, testGroup)
 import Test.Tasty.HUnit (Assertion, assertBool, testCase, (@?=))
-import Test.Tasty.QuickCheck (testProperty)
+import Test.Tasty.QuickCheck
```
Comment (Soupstraw) [inline eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs]:
```suggestion
import Test.Tasty.QuickCheck (testProperty)
```

Comment (Soupstraw) [reply eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs]: Not sure why the imports have changed in this file..

File: libs/cardano-ledger-core/src/Cardano/Ledger/Language.hs [hunk 13]
```diff
@@ -96,7 +129,8 @@ instance IsLanguage 'PlutusV2 where
 toSLanguage :: forall (l :: Language) m. (IsLanguage l, MonadFail m) => Language -> m (SLanguage l)
 toSLanguage lang
   | fromSLanguage thisLanguage == lang = pure thisLanguage
-  | otherwise = fail $ "Plutus language mismatch. Expected " ++ show thisLanguage ++ ", but got: " ++ show lang
+  | otherwise =
+      fail $ "Plutus language mismatch. Expected " ++ show thisLanguage ++ ", but got: " ++ show lang
```
Comment (Soupstraw) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Language.hs]: Unnecessary change

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/MirTransfer.hs (hunk 11 above)
Comment (Soupstraw) [reply eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/MirTransfer.hs:29]: never mind, `ppProtocolVersionL` moved to Core.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 14]
```diff
@@ -561,39 +649,65 @@ encodeLangViews views = encodeMapLen n <> foldMap encPair ascending
...
-  col
-  mxCol =
-    AlonzoPParams ma mb mxBB mxT mxBH kd pd emx a n rho tau d eE pv mnP ada cost price mxTx mxBl mxV col mxCol
-
--- ======================================
-
-instance ToExpr (AlonzoPParamsHKD StrictMaybe era)
-
-instance ToExpr (AlonzoPParamsHKD Identity era)
+-- | Given the missing pieces Turn a ShelleyPParams into a AlonzoPParams
```
Comment (JaredCorduan) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]:
```suggestion
-- | Given the missing pieces, turn a ShelleyPParams into a AlonzoPParams
```

File: eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Examples/Consensus.hs [hunk 15]
```diff
@@ -52,13 +50,12 @@ import qualified Data.Map.Strict as Map
 import Data.Proxy (Proxy (..))
 import qualified Data.Sequence.Strict as StrictSeq
 import qualified Data.Set as Set
+import Lens.Micro
```
Comment (JaredCorduan) [inline eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Examples/Consensus.hs:53]: is this a library that we also want to allow open imports for? :thinking:

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs [hunk 16]
```diff
@@ -3,34 +3,83 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE GADTs #-}
+{-# LANGUAGE InstanceSigs #-}
```
Comment (JaredCorduan) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:6]: :heart:

File: eras/mary/impl/src/Cardano/Ledger/Mary/Translation.hs [hunk 17]
```diff
@@ -117,7 +114,7 @@ instance Crypto c => TranslateEra (MaryEra c) ShelleyTxOut where
   translateEra () (TxOutCompact addr cfval) =
     pure $ TxOutCompact (coerce addr) (translateCompactValue cfval)
 
-instance Crypto c => TranslateEra (MaryEra c) UTxO where
+instance (Crypto c) => TranslateEra (MaryEra c) UTxO where
```
Comment (JaredCorduan) [inline eras/mary/impl/src/Cardano/Ledger/Mary/Translation.hs]: I thought you were the destroyer of extraneous parenthesis :laughing:

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs [hunk 18]
```diff
@@ -449,26 +442,6 @@ validateVerifiedWits tx =
...
-  Tx era ->
-  Set (KeyHash 'Witness (Crypto era)) ->
-  Test (ShelleyUtxowPredFailure era)
-validateNeededWitnesses genDelegs utxo tx witsKeyHashes =
-  let needed = witsVKeyNeeded utxo tx genDelegs
-      missingWitnesses = diffWitHashes needed witsKeyHashes
-   in failureUnless (nullWitHashes missingWitnesses) $
-        MissingVKeyWitnessesUTXOW missingWitnesses
--}
-
```
Comment (JaredCorduan) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs:471]: :pray:

**Review by JaredCorduan — APPROVED** [review]
This all looks amazing, thank you!

You don't have to do this, just an idea: maybe we could add a section to the `CONTRIBUTING.md` doc about open imports, forbidden them except for `Core` modules, and maybe lenses?

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/API/ByronTranslation.hs (hunk 7 above)
Comment (lehins) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/API/ByronTranslation.hs]: Yep, done.

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs (hunk 8 above)
Comment (lehins) [reply eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs]: Thank you! Done

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs (hunk 9 above)
Comment (lehins) [reply eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/GenesisDelegation.hs]: Done

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/Mir.hs (hunk 10 above)
Comment (lehins) [reply eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/Mir.hs]: Good catch. Done

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs (hunk 12 above)
Comment (lehins) [reply eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs]: That's because I also ended up needing `Property`,  `discard`, `(===)`, but I wasn't sure what else I might need, thus ended up removing it temporarily, but forgot to make it explicit again.

Fixed

File: libs/cardano-ledger-core/src/Cardano/Ledger/Language.hs (hunk 13 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Language.hs]: This one actually was need, since a line was way too long. Taking this part to another line shaved off 10 characters :slightly_smiling_face: 
Its is still a bit too long, I'll switch it to a multi-line.

Done

File: eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Examples/Consensus.hs (hunk 15 above)
Comment (lehins) [reply eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Examples/Consensus.hs:53]: Yeah, for sure, since now it became omnipresent.

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs (hunk 16 above)
Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:6]: Yeah. Neat trick. It is need to get to the `f` type parameter in the instance definition:
```haskell
  hkdMaxTxExUnitsL :: forall f. HKDFunctor f => Lens' (PParamsHKD f (ConwayEra c)) (HKD f ExUnits)
  hkdMaxTxExUnitsL =
    lens (hkdMap (Proxy @f) unOrdExUnits . bppMaxTxExUnits) $ \pp x ->
      pp {bppMaxTxExUnits = hkdMap (Proxy @f) OrdExUnits x}
```

File: eras/mary/impl/src/Cardano/Ledger/Mary/Translation.hs (hunk 17 above)
Comment (lehins) [reply eras/mary/impl/src/Cardano/Ledger/Mary/Translation.hs]: I am! I have no idea how I could commit such an atrocity!
:smile: 

Fixed

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 14 above)
Comment (lehins) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: Done
